services/api/src/api/routes/rows.py

```diff
@@ -0,0 +1,319 @@
+        row_group_offsets (Any): The row group offsets. See index().
+        row_group_readers (List[RowGroupReader]): The row group readers. See index().
+
+    Returns:
+        pa.Table: The requested rows.
+    """
+    if (len(row_group_offsets) == 0) or (len(row_group_readers) == 0):
+        raise ParquetResponseEmptyError("No parquet files found.")
+    last_row_in_parquet = row_group_offsets[-1] - 1
+    first_row = min(offset, last_row_in_parquet)
+    last_row = min(offset, offset + length - 1, last_row_in_parquet)
+    first_row_group_id, last_row_group_id = np.searchsorted(row_group_offsets, [first_row, last_row], side="right")
+    pa_table = pa.concat_tables([row_group_readers[i]() for i in range(first_row_group_id, last_row_group_id + 1)])
+    first_row_in_pa_table = row_group_offsets[first_row_group_id - 1] if first_row_group_id > 0 else 0
+    return pa_table.slice(offset - first_row_in_pa_table, length)
+
+
+class FeatureItem(TypedDict):
+    feature_idx: int
+    name: str
+    type: Mapping[str, Any]
+
+
+# in JSON, dicts do not carry any order, so we need to return a list
+#
+# > An object is an *unordered* collection of zero or more name/value pairs, where a name is a string and a value
+#   is a string, number, boolean, null, object, or array.
+# > An array is an *ordered* sequence of zero or more values.
+# > The terms "object" and "array" come from the conventions of JavaScript.
+# from https://stackoverflow.com/a/7214312/7351594 / https://www.rfc-editor.org/rfc/rfc7159.html
+def to_features_list(features: Features) -> List[FeatureItem]:
+    features_dict = features.to_dict()
+    return [
+        {
+            "feature_idx": idx,
+            "name": name,
+            "type": features_dict[name],
+        }
+        for idx, name in enumerate(features)
+    ]
+
+
+class RowItem(TypedDict):
+    row_idx: int
+    row: Mapping[str, Any]
+    truncated_cells: List[str]
+
+
+def to_rows_list(pa_table: pa.Table, offset: int, features: Features, unsupported_columns: List[str]) -> List[RowItem]:
+    num_rows = pa_table.num_rows
+    for idx, (column, feature) in enumerate(features.items()):
+        if column in unsupported_columns:
+            pa_table = pa_table.add_column(idx, column, pa.array([None] * num_rows))
+    return [
+        {
+            "row_idx": idx + offset,
+            "row": row,
+            "truncated_cells": unsupported_columns,
+        }
+        for idx, row in enumerate(pa_table.to_pylist())
+    ]
+
+
+def create_response(pa_table: pa.Table, offset: int, features: Features, unsupported_columns: List[str]) -> Any:
+    if set(pa_table.column_names).intersection(set(unsupported_columns)):
+        raise RuntimeError(
+            "The pyarrow table contains unsupported columns. They should have been ignored in the row group reader."
+        )
+    return {
+        "features": to_features_list(features),
+        "rows": to_rows_list(pa_table, offset, features, unsupported_columns),
```

Comment:
Sorry if I misunderstood but should it be pa_table.column_names instead of unsupported_columns?